rfc/rfc-78/rfc-78.md

```diff
@@ -0,0 +1,220 @@
+of the table to ensure all latest file slices has just the base files. 
+  - Lets dissect and see what it needs to support not requiring the full compaction. In general, we plan to add a table config to track the commit time (more on this later in this doc) when the upgrade was done. 
+    So, using the upgrade commit time, we should be able to use different log file comparator to order log files within a given file slice. 
+
+## 1.0 ➝ 0.16.0 downgrade steps
+Users will have to use hudi-cli to do the downgrade. Here are the steps we might need to do during downgrade.
+- Similar to our upgrade, lets trigger full compaction to latest file slice so that all latest file slices only has base files. We do not want to have any file slices w/ a mix of 1.x log files and 0.x log files.
+- Rewrite LSM based timeline to archived timeline. we have to deduce writer properties and introduce boundaries based on that (i.e. until which commit to archive).
+- We have two options wrt handling active timeline instants:
+   A: No changes to active instants. In order to support 1.x tables w/ 0.16.0 reader, already these changes might have been ported to 0.16.0 reader. But we might have to ensure downgrade from 0.16.0 to any older version of hudi does take care of rewriting the active timeline instants.
+   B: While downgrading from 1.x to 0.16.0, let’s rewrite the active timeline instants too. So that after downgrade, we can't differentiate whether table was natively created in 0.16.0 or was it upgraded to 1.x and then later downgraded. The con in Option A is not an issue here since we take care of active timeline instants during 1.x to 0.16.0 downgrade only. So, downgrade from 0.16.0 to any lower hudi version does not need to do any special handling if table was native to 0.16.0 or was it downgraded from 1.x to 0.16.0
+   Our proposal is to go with Option B to ensure 0.16.0 table will be intact and not have any residues from 1.x writes. While the rewrite of active instants are happening, we could see some read failures. But 
+   we do not want to have a 0.16.0 (post upgrade) table to deal w/ 1.x timeline intricacies. 
+- If there are new MDT partitions (functional index, sec index), nuke them. Update table properties.
+- Check latest snapshot files for log files w/ new features enabled. For eg, deletion vector, partial update. If any log files are found in latest snapshot, we have to trigger compaction for those file groups. This calls for a custom compaction strategy as well which compacts only for certain file groups.
+   - In order to reduce downtime to downgrade, we are inclining to do this only for snapshot reads. If there are older file slices and users are issuing time travel or incremental reads, and if they have any of the new features enabled (deletion vector, partial update), readers could break after downgrade.
+
+## Implementation details
```

Comment:
Could we have a separate sections to discuss how all different query types, i.e., snapshot, read-optimized, incremental, CDC, time-travel, operate correctly before and after migration in different scenarios of reader and writer (0.15 -> 0.16 -> 1.0), e.g., the below scenario of independently upgrading reader and writer?
```
writer:  <--- 0.15 ------>  |  <-------- 0.16 ------> | <-------- 1.0 ----
reader: <--- 0.15 -------------->   | <-------- 0.16 -----------> | <-------- 1.0 ----
```

Reply:
can you clarify me something as to what you want me to cover here. 
we are only adding 1.x read support to 0.16.0. but 1.x should be able to read any tables from 0.14.0 to 1.x. So, we can't really use 0.15.0 reader to read 1.x table. So, users have to upgrade to 0.16.x first before they can read any of 1.x tables.

Reply:
probably just adding a matrix table for 0.16.x and 1.x, with different write and read scenarios.